File: scala/src/main/scala/domain/Bank.scala
package domain

import domain.ConversionResult.{fromFailure, fromSuccess}
import domain.Currency.Currency

sealed case class Bank private (
    private val exchangeRates: Map[String, Double] = Map.empty
) {
  private def keyFor(from: Currency, to: Currency): String = s"$from->$to"

  def addExchangeRate(from: Currency, to: Currency, rate: Double): Bank =
    Bank(exchangeRates.updated(keyFor(from, to), rate))

  def convert(money: Money, toCurrency: Currency): ConversionResult[String] = {
    if (canConvert(money, toCurrency))
      fromSuccess(convertSafely(money, toCurrency))
    else fromFailure(s"${money.currency}->$toCurrency")
  }

  private def canConvert(money: Money, toCurrency: Currency): Boolean =
    money.currency == toCurrency || exchangeRates.contains(
      keyFor(money.currency, toCurrency)
    )

  private def convertSafely(money: Money, toCurrency: Currency): Money =
    if (money.currency == toCurrency) money
    else
      Money(
        money.amount * exchangeRates(keyFor(money.currency, toCurrency)),
        toCurrency
      )
}

object Bank {
  def withExchangeRate(from: Currency, to: Currency, rate: Double): Bank =
    Bank().addExchangeRate(from, to, rate)
}

File: c#/money-problem.Domain/Bank.cs
namespace money_problem.Domain
{
    public sealed class Bank
    {
        private readonly Dictionary<string, double> _exchangeRates;

        private Bank(Dictionary<string, double> exchangeRates) => this._exchangeRates = exchangeRates;

        public static Bank WithExchangeRate(Currency from, Currency to, double rate)
        {
            var bank = new Bank(new Dictionary<string, double>());
            bank.AddExchangeRate(from, to, rate);
            return bank;
        }

        public void AddExchangeRate(Currency from, Currency to, double rate)
            => this._exchangeRates[KeyFor(from, to)] = rate;

        private static string KeyFor(Currency from, Currency to) => $"{from}->{to}";

        public Money Convert(Money money, Currency to) =>
            this.CanConvert(money.Currency, to)
                ? this.ConvertSafely(money, to)
                : throw new MissingExchangeRateException(money.Currency, to);

        private Money ConvertSafely(Money money, Currency to) =>
            to == money.Currency
                ? money
                : new Money(money.Amount * this._exchangeRates[KeyFor(money.Currency, to)], to);

        private bool CanConvert(Currency from, Currency to) =>
            from == to || this._exchangeRates.ContainsKey(KeyFor(from, to));
    }
}